### ISSUE TITLE
[Bug] Reused Prior instances are quietly ignored

### ISSUE THREAD

**Issue by mrcslws** [cornellius-gp/gpytorch#2263]
# 🐛 Bug

If a model uses a single `Prior` instance for multiple registered priors, only one of them will have any effect.

For example, if a user adds a `lengthscale_prior` argument to botorch's [MixedSingleTaskGP](https://github.com/pytorch/botorch/blob/b3d3074e7f0a8142a86ec5406b4f6553aba4e769/botorch/models/gp_regression_mixed.py#L28) and passes this argument down to the four kernels, gpytorch will quietly ignore the prior for three of those kernels.

This happens silently, with no warning or error. It took me months to notice that there was an issue, and it took a lot of in-depth debugging to track it down.

## To reproduce

```python
import copy

import gpytorch
import torch

def my_additive_kernel(lengthscale_prior, copy_the_prior=False):
    kernel1 = gpytorch.kernels.MaternKernel(
        ard_num_dims=2,
        active_dims=[0, 1],
        lengthscale_prior=lengthscale_prior
    )

    if copy_the_prior:
        # Workaround
        lengthscale_prior = copy.deepcopy(lengthscale_prior)

    kernel2 = gpytorch.kernels.MaternKernel(
        ard_num_dims=2,
        active_dims=[2, 3],
        lengthscale_prior=lengthscale_prior
    )

    return kernel1 + kernel2


class MyModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood, covar_module):
        super().__init__(train_x, train_y, likelihood)
        self.covar_module = covar_module


def print_toy_mll(covar_module):
    likelihood = gpytorch.likelihoods.GaussianLikelihood()
    train_x = torch.tensor([])
    train_y = torch.tensor([])
    model = MyModel(
        train_x, train_y, likelihood,
        covar_module=covar_module,
    )
    mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

    # Fake model output; we only care about the Prior logic
    function_dist = gpytorch.distributions.MultivariateNormal(torch.tensor([0.0]),
                                                              torch.tensor([[1.0]]))
    target = torch.tensor([0.0])
    with torch.no_grad():
        print(mll(function_dist, target))


print("Reusing same prior instances, mll is:")
print_toy_mll(covar_module=my_additive_kernel(
    lengthscale_prior=gpytorch.priors.GammaPrior(3.0, 6.0),
))
print("Using unique prior instances, mll is:")
print_toy_mll(covar_module=my_additive_kernel(
    lengthscale_prior=gpytorch.priors.GammaPrior(3.0, 6.0),
    copy_the_prior=True,
))

```

**Output**
```
Reusing same prior instances, mll is:
tensor(-1.6018)
Using unique prior instances, mll is:
tensor(-2.0214)
```

## Expected Behavior

The code's behavior should not depend on whether priors are reused. Or the code should throw an error when a model has a single instance registered multiple times.

Conceptually, a Prior is just an immutable object, similar to a torch.distributions instance, so it feels strange/unexpected that gpytorch cares whether two parameters are sharing prior instances.

## System information

**Please complete the following information:**
gpytorch: 1.9.1.dev32+g23b068b5
torch: 1.13.1
Mac OS 13.2 M1

## Extra context

This bug is due to the memory logic in `module.named_priors()`. 

https://github.com/cornellius-gp/gpytorch/blob/2e77f124cf78d7adb3ee2fb3bc97d74a7de0a9c9/gpytorch/module.py#L531

Either that method should be fixed, or the MLLs should stop using it. Personally, I think `named_priors` is broken; if `register_prior` is called two times, then `named_priors` should return two names, regardless of whether the two calls used the same instance.

**Comment by Balandat** [comment]
Thanks for the bug report. I believe this is b/c of how we memoize the priors internally here when getting the priors of a module: https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/module.py#L526-L540

I think using the same prior on different (sub)modules just is a use case that we haven't really considered; it seems like it should be relatively straightforward to achieve the correct behavior here, let me try and take a stab.

**Comment by Balandat** [comment]
Oh I see you already flagged this in the extra context, thanks.

**Comment by mrcslws** [comment]
Confirmed fixed in v1.10. 👍
